Rewrite service checks

murrant opened the pull request:
Rewrite services to improve security
Adds a services polling module, but still polls through the separate check-services.php or lnms services:poll scripts by default.
Attempts to parse help text into valid options.
Default values for options such as -H.

TODO:

- [ ] Improve help parsing and tests
- [x] Test polling
- [ ] Test all the webui
- [x] Convert all old check scripts.

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

murrant committed:
Rewrite Sensors code
Remove includes/services.inc.php

murrant committed:
fix check-services (replace with services:poll)

murrant committed:
Tidy up things with a DefaultServiceCheck
Implement parsing of check parameters for use in the webui.

murrant committed:
Starting on webui

murrant committed:
working paramaters

murrant committed:
Saving working

murrant committed:
import toggle component from the other branch

murrant committed:
working param error display

murrant committed:
translation

murrant committed:
help parsing fix, inclusive groups support, test

murrant committed:
improved validation

murrant committed:
clean up some items

murrant committed:
escape more correctly

murrant committed:
test the service check

murrant committed:
port over add services pages

murrant committed:
initial service page port

murrant committed:
small fixes, modal version firing twice??

murrant committed:
Fix polling, adding, testing, and more

murrant committed:
delete working

murrant committed:
restore body-class attribute...  not sure how to pass attributes to the default slot

murrant committed:
editing service

murrant committed:
All services page ported

murrant committed:
misc fixes

murrant committed:
Change how default values work

murrant committed:
Default Params WIP

murrant committed:
Colored service print

murrant committed:
module cleanup a bit

murrant committed:
check-services is deprecated

murrant committed:
prevent check-services and services:poll if polling module is enabled.  Add ./poller.php shim

murrant committed:
rebase, regenerate css and remove double escape
Fix submitCheck method logic once and for all

murrant committed:
Update links to use routes, remove unused, add basic view, add filters for disabled and ignored

murrant committed:
style and lint fixes

murrant committed:
change visibility

murrant committed:
Migrate existing checks and update graphs

moschroe commented:
As suggested on discord, here my proposed extensions for service plugins:

1. It ought to be possible to keep the original name of perf metrics for the graph and use the canonicalised form only for interfacing with RRD, allowing spaces and punctuation in the legend.
2. Also, there should be an option available to pass SNMP credentials (+ protocol, sec level, etc.) via envvar to the service script. This would allow using existing SNMP facilities (extend scripts, forwarded ports, firewall rules) while not increasing exposed surface.

I guess the second point would be more straightforward to implement. I might be able to contribute there, let me know whether that would be desirable or I should open a separate PR (probably waiting on this one then)

murrant commented:
@moschroe Thanks for the input.  If you would like to help out with this, you can send a pull request directly to my branch. https://github.com/murrant/librenms/tree/rewrite_services

Also, creating services is quite broken right now in this PR.

murrant committed:
Style fixes

murrant committed:
Fix adding services and improve test output

murrant committed:
Change rrd files to one ds per rrd
This allows dynamic datasets without breaking graphs.
Add the ability to import existing data to make migration painless.

murrant committed:
remove legacy data file updates

murrant committed:
Better new file names
fix bug in old ds names
docs and misc improvements

murrant committed:
consistent Controller name

murrant committed:
disabled color service when disabled

murrant committed:
fix missing uom

murrant committed:
Fix rrd source error when missing

murrant committed:
update service_ds to empty if appropriate

murrant committed:
Show cli for unknown

murrant committed:
Don't set -H for checks that don't have -H
Set service_ip to null to signify this.

murrant committed:
consolidate service validation

murrant committed:
improve help parser to handle groups with options after

murrant committed:
Better visibility for parameter descriptions

murrant committed:
avoid double / for a cleaner look

murrant committed:
check_disk parsing WIP

murrant committed:
Don't destroy legacy params

murrant committed:
Move getMetrics to the DefaultServiceCheck so it can be overridden
Work around upstream mssql_health bug where KB is kb.

murrant committed:
quotes around the ds

murrant committed:
fix silly surrounding quotes

murrant committed:
Fix exclusive groups

murrant committed:
service_ip should be null if --hostname does not exist

murrant committed:
Convert legacy params on edit
fix param loading

murrant committed:
Make defaults actually work correctly and more conveniently

murrant committed:
hook up param validation rules for override in service check

murrant committed:
cleanups

murrant committed:
prevent parameters from being wiped when loading a service to edit
(but allow it in all other situations)

murrant committed:
Update module to new interface
Ability to dump, but no way to feed data right now.  Could mock check cli output in the future.

murrant committed:
Merge params correctly
Display placeholder correctly

murrant committed:
Only find usage once

murrant committed:
don't match alternate long opts

murrant committed:
update test

murrant committed:
Add help parser tool I made
Parse the correct line

murrant committed:
space

murrant commented:
Needs rework due to not including arguments, only options.
